### PR TITLE
[WIP] initial stab at implementing EIP2718

### PR DESCRIPTION
**Pull request by victor-wei126**
This is a very basic preliminary attempt at EIP2718, mostly to get my thoughts down.

### CODE REVIEW

**Comment by synlestidae** [comment]
So you think the library should implement EIP-1559 and EIP-2930? Any other transaction types that would be "must have"?

Good to see this work. I am working in `improvement/eip-2718/basics`. I think a Transaction trait will play nicely. I plan to use use default trait methods to avoid repeating code. Once these basics are finalised, you should be able to integrate them into your branch yeah? Hoping to have it done by end of week.

**Comment by victor-wei126** [comment]
I _believe_ EIP 1559 and 2390 are the only other "officially" supported transaction types currently, besides the basic legacy transactions that we already have. There was another EIP about batch transactions, but that was withdrawn. So I think at the minimum these are good, and we should design our library so that is future-proof and easy to support new transactions as they come up.

Out of curiosity what made you decide to use a Trait? I remember this was brought up but you had mentioned a trait might not be necessary. Taking a brief look at your code, I think this could work out. I can probably implement the trait on the enum I've defined.

**Comment by synlestidae** [comment]
I decided to use a trait so that, as you say, a new transaction can be supported without it being a breaking change. When you add or remove from an enum in rust, this is a breaking change.

When there is a new transaction type, add the struct and support the trait
e.g.

```rust
impl Transaction for EIP9999Txn {
//...
} 
```

File: src/raw_transaction.rs [hunk 1]
```diff
@@ -125,6 +185,15 @@ mod test {
     use std::io::Read;
     use ethereum_types::H256;
 
+    #[test]
+    fn test_eip2718() {
+        //let txn = Transaction::FeeMarketEIP1559(EIP1559Txn::new(...));
+        let txn = EIP1559Txn::new(...);
+        let txn2 = EIP2930Txn::new(...);
+        let txn_3 = LegacyTxn::new(...);
+        let bytes = txn.sign(secret_key);
```
Comment (victor-wei126) [inline src/raw_transaction.rs:194]: @synlestidae This is what I'm thinking for the user-facing API, just two simple methods, new() and sign(). New gives the user the ability to instantiate a new (raw) transaction. Then, they can sign the transaction, which under the hood handles all the encoding and signing, and returns the byte-encoded transaction. What are your thoughts?

**Comment by victor-wei126** [comment]
@synlestidae Sorry about the delay in implementation, a lot of things have been going on on my end. I will try to get an update up this weekend and start at it again this coming week.

**Comment by synlestidae** [comment]
Not a problem @victor-wei126, there is no obligation. I am working on a branch on EIP-2930 which you can see under #41.

**Comment by victor-wei126** [comment]
I'm going to close this PR since it seems that it is out of date and you already started work on 2930 based transactions.

**Pull request closed without merging**